codex/merkletree/merkletree.nim

```diff
@@ -0,0 +1,111 @@
+import std/sequtils
+import std/math
+import std/bitops
+import std/sugar
+
+import pkg/libp2p
+import pkg/stew/byteutils
+
+import pkg/questionable
+import pkg/questionable/results
+
+
+type
+  MerkleHash* = MultiHash
+  MerkleTree* = ref object of RootObj
+    nodes: seq[seq[MerkleHash]]
+  MerkleTreeBuilder* = ref object of RootObj
+    mcodec: ?MultiCodec
+    leaves: seq[MerkleHash]
+  MerkleProof* = ref object of RootObj
+    path: seq[MerkleHash]
+
+proc `$`*(t: MerkleTree): string =
+  result &= "height: " & $t.nodes.len
+  result &= "\nnodes: " & $t.nodes
+
+proc addLeaf*(self: MerkleTreeBuilder, hash: MerkleHash): ?!void =
+  if codec =? self.mcodec:
+    if codec != hash.mcodec:
+      return failure("Expected codec is " & $codec & " but " & $hash.mcodec & " received")
+  else:
+    self.mcodec = hash.mcodec.some
+
+  self.leaves.add(hash)
+  return success()
+
+func calcTreeHeight(leavesCount: int): int =
+  if isPowerOfTwo(leavesCount): 
+    fastLog2(leavesCount) + 1
+  else:
+    fastLog2(leavesCount) + 2
+
+proc build*(self: MerkleTreeBuilder): ?!MerkleTree =
```

Comment:
Definitely bike shedding here, but the whole `mybuilder.build()` always seemed redundant to me :sweat_smile: Now `mybuilder.toTree()` :sunglasses:

```suggestion
proc toTree*(self: MerkleTreeBuilder): ?!MerkleTree =
```

Ok I'll stop bikes shedding for now. ;)

Reply:
Builder got replaced with a single `newTree` proc.

Reply:
I really recommend taking a look at the [status style guide](https://status-im.github.io/nim-style-guide/). 

The tl;dr is that you want to use `new` to construct `ref` types and `init` to construct objects, since we want to make `MerkleTree` an object instead of a reference, we should be using `init`.

Reply:
Thanks for the info. I used `init` as suggested.